<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width,initial-scale=1.0">
  <title>Promises</title>
  <link rel="stylesheet" href="../base.css">
</head>

<body>
  <script>
    function makePizza(toppings = []) {
      return new Promise(function(resolve, reject) {
        if(toppings.includes('pineapple')) {
          reject('get out 🍍');
        }
        const amountOfTimetoBake = 500 + (toppings.length * 200);
        // wait one second for the pizza to cook
        setTimeout(function () {
          // when you are ready, you can resolve this promise
          resolve(`Here is your 🍕 with the toppings: ${toppings.join(', ')}`);
        }, amountOfTimetoBake);
        // if something went wrong, you can reject this promise
      });
    }

    // const pepperoniPromise = makePizza(['pepperoni']);
    // const canadianPromise = makePizza(['pepperoni', 'mushrooms', 'onions']);

    // pepperoniPromise.then(function(pizza) {
    //   console.log('Ah, got it.');
    //   console.log(pizza);
    // })

    // console.log(pepperoniPromise);
    // console.log(canadianPromise);

    makePizza(['pepperoni'])
      .then(function (pizza) {
        console.log(pizza);
        return makePizza(['pineapple']);
      })
      .then(function (pizza) {
        console.log(pizza);
        return makePizza(['cheese', 'shrooms', 'feta', 'onion', 'garlic', 'ham']);
      })
      .then(function (pizza) {
        console.log(pizza);
      })
      .catch(handleError);
      // it catches the error and stops the chain, so the other pizzas also sdont get made

      // run them concurrently
      const pizzaPromise1 = makePizza(['bellpepper', 'feta', 'garlic'])
      const pizzaPromise2 = makePizza(['eggplant', 'cheese', 'shrooms', 'feta', 'onion', 'garlic', 'ham'])
      const pizzaPromise3 = makePizza(['ham', 'cheese'])

      const dinnerPromise = Promise.all([pizzaPromise1, pizzaPromise2, pizzaPromise3]);

      // dinnerPromise.then(function(pizzas) {
      //   const [hottie, garbagePail, hamAndCheese] = pizzas;
      //   console.log(hottie, garbagePail, hamAndCheese)
      // })

      dinnerPromise.then(function([hottie, garbagePail, hamAndCheese]) {
        console.log(hottie, garbagePail, hamAndCheese)
      })

      const firstPizzaPromise = Promise.race([pizzaPromise1, pizzaPromise2, pizzaPromise3]);

      firstPizzaPromise.then(pizza => {
        console.log('you must be hungry');
        console.log(pizza);
      })

      function handleError(err) {
        console.log('on noes');
        console.log(err);
      }

      // a then will run when a promise is successfull, catch when it isnt
      makePizza(['cheese', 'pineapple'])
      .then(pizza => {
        console.log(pizza);
      }).catch(handleError)

      const p1 = makePizza(['pep']);
      const p2 = makePizza(['pineapple'])

      const dinnerPromise2 = Promise.allSettled([p1, p2]);

      dinnerPromise2.then(results => {
        console.log(results);
      })

  </script>
</body>

</html>
